torchvision/prototype/transforms/_auto_augment.py

```diff
@@ -43,29 +32,31 @@ def _get_random_item(self, dct: Dict[K, V]) -> Tuple[K, V]:
         key = keys[int(torch.randint(len(keys), ()))]
         return key, dct[key]
 
-    def _extract_image(
+    def _flatten(
         self,
         sample: Any,
         unsupported_types: Tuple[Type, ...] = (features.BoundingBox, features.SegmentationMask),
-    ) -> Tuple[Tuple[Any, ...], Union[PIL.Image.Image, torch.Tensor, features.Image]]:
-        def fn(
-            id: Tuple[Any, ...], inpt: Any
-        ) -> Optional[Tuple[Tuple[Any, ...], Union[PIL.Image.Image, torch.Tensor, features.Image]]]:
-            if type(inpt) in {torch.Tensor, features.Image} or isinstance(inpt, PIL.Image.Image):
-                return id, inpt
+    ) -> Tuple[List[Any], TreeSpec, int, Union[PIL.Image.Image, torch.Tensor, features.Image]]:
+        sample_flat, sample_spec = tree_flatten(sample)
+        images = []
+        for idx_flat, inpt in enumerate(sample_flat):
+            if isinstance(inpt, (features.Image, PIL.Image.Image)) or is_simple_tensor(inpt):
+                images.append((idx_flat, inpt))
             elif isinstance(inpt, unsupported_types):
                 raise TypeError(f"Inputs of type {type(inpt).__name__} are not supported by {type(self).__name__}()")
-            else:
-                return None
 
-        images = list(query_recursively(fn, sample))
         if not images:
             raise TypeError("Found no image in the sample.")
         if len(images) > 1:
             raise TypeError(
                 f"Auto augment transformations are only properly defined for a single image, but found {len(images)}."
             )
-        return images[0]
+        idx_flat, image = images[0]
+        return sample_flat, sample_spec, idx_flat, image
```

Comment:
The first three parameters are never touched by the transforms, but are needed to put the new image in the right place before we unflatten later. Should we wrap them in a `dataclass` or something? Or maybe just in a tuple like

```suggestion
        return (sample_flat, sample_spec, idx_flat), image
```

to indicate their cohesion?

Reply:
Its an internal method, so any convention is ok for me

Reply:
I agree with Victor. It's best if we keep things as simple as possible. For me it's still unclear why we need to keep the 3 items and not just return as previously the image and its location and append it at the end. I'm always supportive when we reuse methods from core and reduce the utility methods on our side but in this case it seems that we are complicating things. I wonder if it's possible to still reuse parts of pytree but keep our implementation closer to the "extract image" logic.

Reply:
> For me it's still unclear why we need to keep the 3 items and not just return as previously the image and its location and append it at the end. 

We can't append it, but rather put it in the right place. If the input is not a flat dict or list, we need more than one index / key. The `query_recursively` gave us and `id` which is this collection of indices and keys. In the new implementation the equivalent is `idx_flat`. 

The difference is that the `id` of the previous implementation corresponds to the original sample, whereas `idx_flat` corresponds to the flattened sample.

Reply:
Thanks for the explanation. That makes sense.

Does it make sense not to replace our previous implementation or to replace smaller components of it? It seems what we did previously was both efficient and easier to read. Here we try to reduce the LOC (which I always like) but seems that this isn't for free both in terms of speed and readability.